pkg/mutation/schema/schema.go

```diff
@@ -0,0 +1,371 @@
+func (db *DB) Upsert(mutator MutatorWithSchema) error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	return db.upsert(mutator, true)
+}
+
+func (db *DB) upsert(mutator MutatorWithSchema, unwind bool) error {
+	oldMutator, ok := db.mutators[mutator.ID()]
+	if ok && !oldMutator.HasDiff(mutator) {
+		return nil
+	}
+	if ok {
+		db.remove(oldMutator.ID())
+	}
+	bindings := mutator.SchemaBindings()
+	gvks := map[schema.GroupVersionKind]struct{}{}
+	// deduplicate GVKs
+	for _, binding := range bindings {
+		for _, group := range binding.Groups {
+			for _, version := range binding.Versions {
+				for _, kind := range binding.Kinds {
+					gvk := schema.GroupVersionKind{
+						Group:   group,
+						Version: version,
+						Kind:    kind,
+					}
+					gvks[gvk] = struct{}{}
+				}
+			}
+		}
+	}
+
+	modified := []*scheme{}
+	for gvk := range gvks {
+		s, ok := db.schemas[gvk]
+		if !ok {
+			s = &scheme{gvk: gvk}
+			db.schemas[gvk] = s
+		}
+		if err := s.add(mutator.Path().Nodes); err != nil {
+			// avoid infinite recursion
+			if unwind {
+				db.unwind(mutator, oldMutator, modified)
+			}
+			return err
+		}
+		modified = append(modified, s)
+	}
+	m := mutator.DeepCopy()
+	db.mutators[mutator.ID()] = m.(MutatorWithSchema)
+	return nil
+}
+
+// unwind a bad commit
+func (db *DB) unwind(new, old MutatorWithSchema, schemes []*scheme) {
+	for _, s := range schemes {
+		s.remove(new.Path().Nodes)
+		if s.root == nil {
+			delete(db.schemas, s.gvk)
+		}
+	}
+	if old == nil {
+		return
+	}
+	if err := db.upsert(old, false); err != nil {
+		// We removed all changes made by the previous mutator and
+		// are re-adding a mutator that was already present. Because
+		// this mutator is already present and we have a lock on the
+		// db, this should never fail. If it does we are in an unknown
+		// state and should panic so we can recover by bootstrapping
+		// and raise the visibility of the issue.
+		log.Error(err, "could not upsert previously existing mutator into schema, this is not recoverable")
+		panic(err)
+	}
+}
+
+// Remove removes the mutator with the given id from the
+// db.
+func (db *DB) Remove(id types.ID) {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	db.remove(id)
+	delete(db.mutators, id)
+}
+
+func (db *DB) remove(id types.ID) {
+	mutator, ok := db.mutators[id]
+	if !ok {
+		// no mutator found, nothing to do
+		return
+	}
+	bindings := mutator.SchemaBindings()
+	gvks := map[schema.GroupVersionKind]struct{}{}
+	// deduplicate GVKs
+	for _, binding := range bindings {
+		for _, group := range binding.Groups {
+			for _, version := range binding.Versions {
+				for _, kind := range binding.Kinds {
+					gvk := schema.GroupVersionKind{
+						Group:   group,
+						Version: version,
+						Kind:    kind,
+					}
+					gvks[gvk] = struct{}{}
+				}
+			}
+		}
+	}
+
+	for gvk := range gvks {
+		s, ok := db.schemas[gvk]
+		if !ok {
+			log.Error(nil, "mutator associated with missing schema", "mutator", id, "schema", gvk)
```

Comment:
Should we panic here too? Even though the state is consistent, nobody will see this error otherwise